src/polly/src/event_manager.rs

```diff
@@ -246,354 +157,335 @@ impl EventManager {
-                Some(evset) => evset,
-                None => {
-                    // Ignore unknown bits in event mask.
-                    // TODO: We might want to log this as a warning but that requires a logger
-                    // reference and a crate dependency.
-                    continue;
-                }
-            };
-
-            if let Some(event_handler_data) = self.handlers.get(event_data as i32) {
-                if let Ok(()) = self.dispatch_event(
-                    event_handler_data.data.0,
-                    evset,
-                    event_handler_data.handler,
-                ) {
-                    continue;
-                } else {
-                    // We might get errors related to PollableOps.
-                    // Need to decide what to do with them.
-                    // Options:
-                    // 1. Break loop and throw them to the caller
-                    // 2. Invoke a TBD event handler error callback.
-                }
-            } else {
-                // There is no handler registered for this event set/pollable.
-            }
-        }
-
-        Ok(())
-    }
-
-    // Wait for events, then dispatch to registered event handlers.
+    /// Wait for events, then dispatch to the registered event handlers.
     pub fn run(&mut self) -> Result<usize> {
-        self.run_timeout(-1)
+        self.run_with_timeout(-1)
     }
 
-    // Wait for events or a timeout, then dispatch to registered event handlers.
-    pub fn run_timeout(&mut self, milliseconds: i32) -> Result<usize> {
+    /// Wait for events for a maximum timeout of `miliseconds`. Dispatch the events to the
+    /// registered signal handlers.
+    pub fn run_with_timeout(&mut self, milliseconds: i32) -> Result<usize> {
         let event_count = self
             .epoll
-            .wait(EVENT_BUFFER_SIZE, milliseconds, &mut self.ready_events[..])
+            .wait(
+                EventManager::EVENT_BUFFER_SIZE,
+                milliseconds,
+                &mut self.ready_events[..],
+            )
             .map_err(Error::Poll)?;
-        self.process_events(event_count)?;
+        self.dispatch_events(event_count)?;
 
         Ok(event_count)
     }
-}
 
-// Cascaded epoll support.
-impl EventHandler for EventManager {
-    fn handle_read(&mut self, _source: Pollable) -> Vec<PollableOp> {
-        match self.run_timeout(DEFAULT_EPOLL_TIMEOUT) {
-            Ok(_) => vec![],
-            Err(_) => vec![],
+    fn dispatch_events(&mut self, event_count: usize) -> Result<()> {
+        // Use the temporary, pre-allocated buffer to check ready events.
+        for ev_index in 0..event_count {
+            let event = self.ready_events[ev_index];
+
+            self.subscribers
+                .get_mut(&(event.data() as i32))
+                .unwrap()
+                .clone()
+                .lock()
+                .unwrap()
```

Comment:
@acatangiu @alexandruag @sandreim I wanted to get your opinion on something. `dispatch_events` panics when it cannot find a subscriber associated with the fd for which it has a ready event. This should not happen during normal operation. In case it does, it is a programming error. For example this could happen if we have a bug in `unregister` and we don't properly call `epoll_ctl` to remove the event and the event become ready one way or another. Since this is a programming error, I decided to unwrap, but I just wanted to: 1. call it out. 2. get your input on it.

Reply:
In case we want to keep the existing behavior, I will remove the `Result` return from `dispatch_events`.

Reply:
Had a chat with @raduiliescu and realized that there are some usecases where this could panic without a programatic error. I added a check for the subscriber and for now I'm ignoring the error. We should though consider to add a dependency on a generic logger so that we can log it.

Reply:
I think we have 2 options here:

- log the errors (requires a dependency on a logger)
- propagate the error to event_manager.run() caller context where it can be handled/logged.

Reply:
Propagating the error to the `run` function will break the `run` loop and we will drop events. For level triggered events, that's not such a big problem, but for edge triggered I think we should be careful in dropping unprocessed events. This is why I would propose to just log an error here. As for the dependency on a logger, I don't have such a big concern. I would not use a dependency on our logger implementation, but just on the generic Rust `log` crate. This way we don't make `polly` dependent on Firecracker specific implementation, but we can also have logs for noticing when something is wrong.